src/editor/controls/FlyControls.js

```diff
@@ -1,57 +1,104 @@
-    this.maxXRotation = THREE.Math.degToRad(70);
-  }
 
-  onClickCanvas = () => {
-    if (this.enabled && document.pointerLockElement !== this.inputManager.canvas) {
-      this.inputManager.canvas.requestPointerLock();
-    }
-  };
+    this.onMouseUp = e => {
+      if (this.enabled && e.button === 2) {
+        if (document.pointerLockElement === this.inputManager.canvas) {
+          document.exitPointerLock();
+        }
+      }
+    };
+  }
 
   enable() {
     this.enabled = true;
+    this.inputManager.canvas.requestPointerLock();
     this.inputManager.enableInputMapping(Fly, FlyMapping);
+    this.inputManager.canvas.addEventListener("mouseup", this.onMouseUp);
   }
 
   disable() {
     this.enabled = false;
     this.inputManager.disableInputMapping(Fly);
+    this.inputManager.canvas.removeEventListener("mouseup", this.onMouseUp);
   }
 
   update(dt) {
     if (!this.enabled) return;
-
     const input = this.inputManager;
 
-    const rotation = this.camera.rotation;
-    const lookSensitivity = this.lookSensitivity;
+    // assume that this.camera[position,quaterion/rotation,scale] are authority
+    this.camera.updateMatrix();
+    this.camera.updateMatrixWorld();
+    this.camera.matrixWorld.decompose(worldPos, worldQuat, worldScale);
+
+    // rotate about the camera's local x axis
+    candidateWorldQuat.multiplyQuaternions(
+      quat.setFromAxisAngle(vec.set(1, 0, 0).applyQuaternion(worldQuat), input.get(Fly.lookY) * this.lookSensitivity),
+      worldQuat
+    );
 
-    let xRotation = rotation.x > Math.PI ? rotation.x - 2 * Math.PI : rotation.x;
-    xRotation = THREE.Math.clamp(xRotation, -this.maxXRotation, this.maxXRotation);
+    // check change of local "forward" and "up" to disallow flipping
+    const camUpY = vec.set(0, 1, 0).applyQuaternion(worldQuat).y;
+    const newCamUpY = vec.set(0, 1, 0).applyQuaternion(candidateWorldQuat).y;
+    const newCamForwardY = vec.set(0, 0, -1).applyQuaternion(candidateWorldQuat).y;
+    const extrema = Math.sin(this.maxXRotation);
+    const allowRotationInX =
+      newCamUpY > 0 && ((newCamForwardY < extrema && newCamForwardY > -extrema) || newCamUpY > camUpY);
+    if (allowRotationInX) {
+      this.camera.matrixWorld.compose(
+        worldPos,
+        candidateWorldQuat,
+        worldScale
+      );
+      // assume that if camera.parent exists, its matrixWorld is up to date
+      parentInverse.getInverse(this.camera.parent ? this.camera.parent.matrixWorld : IDENTITY);
+      this.camera.matrix.multiplyMatrices(parentInverse, this.camera.matrixWorld);
+      this.camera.matrixWorld.decompose(this.camera.position, this.camera.quaternion, this.camera.scale);
+    }
 
-    rotation.set(
-      xRotation + input.get(Fly.lookY) * lookSensitivity,
-      rotation.y + input.get(Fly.lookX) * lookSensitivity,
-      0,
-      "ZYX"
+    this.camera.matrixWorld.decompose(worldPos, worldQuat, worldScale);
+    // rotate about the world y axis
```

Comment:
this comment is a little misleading in that we aren't rotating the whole camera transform about the UP vector because we don't want it to translate. We are rotating it "in place" as if we translated the camera to the origin, rotated, then translated back.